scipy/stats/_stats_py.py

```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
+
+        sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+
+        ssbn = 0
+        for sample in samples:
+            smo_ss = _square_of_sums(sample - offset, axis=axis)
+            ssbn = ssbn + smo_ss / sample.shape[axis]
+
+        # Naming: variables ending in bn/b are for "between treatments", wn/w are
+        # for "within treatments"
+        ssbn = ssbn - normalized_ss
+        sswn = sstot - ssbn
+        dfbn = num_groups - 1
+        dfwn = bign - num_groups
+        msb = ssbn / dfbn
+        msw = sswn / dfwn
+        with np.errstate(divide='ignore', invalid='ignore'):
+            f = msb / msw
 
-    sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+        prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
 
-    ssbn = 0
-    for sample in samples:
-        smo_ss = _square_of_sums(sample - offset, axis=axis)
-        ssbn = ssbn + smo_ss / sample.shape[axis]
-
-    # Naming: variables ending in bn/b are for "between treatments", wn/w are
-    # for "within treatments"
-    ssbn = ssbn - normalized_ss
-    sswn = sstot - ssbn
-    dfbn = num_groups - 1
-    dfwn = bign - num_groups
-    msb = ssbn / dfbn
-    msw = sswn / dfwn
-    with np.errstate(divide='ignore', invalid='ignore'):
-        f = msb / msw
+    # if population variances are not equeal, perform Welch ANOVA
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        samples_arrs = np.array(samples)
+
+        # adjust the shape of the data based on the arrary dimension for the calculation
+        # along with input axis
+        if samples_arrs.ndim <= 2:
+            samples_arrs = samples_arrs.T
+        else:
+            samples_arrs = samples_arrs.transpose(
+                *range(1, samples_arrs.ndim - 1), 0, -1
+            )
+
+        means = samples_arrs.mean(axis=axis)
+        ns = np.apply_along_axis(np.size, axis=axis, arr=samples_arrs)
+        k = np.apply_along_axis(np.shape, axis=axis, arr=ns)
+
+        vars_ = np.var(samples_arrs, axis=axis, ddof=1)
+        # calculate weight by number of data and varianc
+        ws = ns / vars_
+
+        # calculate adjusted grand mean
+        mean_prime = (np.sum(ws * means, axis=axis, keepdims=True) /
+                      np.sum(ws, axis=axis, keepdims=True))
+
+        numerator = (np.sum(ws * (means - mean_prime)**2, axis=axis, keepdims=True)
+                     / (k - 1))
+        denominator = (
+                1 + 2 * (k - 2) / (k ** 2 - 1) *
+                np.sum((1 / (ns - 1)) *
+                       (1 - ws / np.sum(ws)) **2,
+                       axis=axis, keepdims=True)
+        )
+        f = numerator / denominator
+
+        # adjusted degree of freedom
+        df = (
+                (k ** 2 - 1) /
+                (3 * np.sum((1 / (ns - 1)) *
+                            (1 - ws / np.sum(ws, axis=axis, keepdims=True)) **2,
```

Comment:
Surprisingly, our linter is not enforcing these rules right now, but:
```suggestion
                            (1 - ws / np.sum(ws, axis=axis, keepdims=True))**2,
```
PEP8 would require equal numbers of spaces before and after, and it also recommends eliminating the space for operators of high priority. Exponentiation has the highest priority, so omit the spaces, please (here and throughout).